[aotd] Subclasses profile logging

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136478
* #139095
* #139068

The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

IvanKobzarev committed:
[aotd] Subclasses profile logging

[ghstack-poisoned]

ezyang commented on torch/_functorch/config.py:
```diff
@@ -201,3 +201,5 @@
 # adds patch, save_config, invalid config checks, etc
 install_config_module(sys.modules[__name__])
+
+aotd_debug_profile: bool = os.environ.get("PYTORCH_AOTD_DEBUG_PROFILE", "0") == "1"
```
needs docs

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"


The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -5989,6 +5990,222 @@ def forward(self, x):
+            )
+
+        # Uncomment for adhoc aotd overhead perf experimentation
+        _print_stat("   SUBCLASS")
```
Is this meant to be commented out before you check it in (So we don't print during CI runs by default)?

Also, I think you mentioned in our 1:1 - but were you thinking of adding this into a benchmark somewhere that we can track over time for regressions? (e.g. we have some benchmarks here, although they are mainly compile-time related https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo/pr_time_benchmarks/benchmarks)

bdhirsh commented on torch/_dynamo/utils.py:
```diff
@@ -887,10 +887,21 @@ def __init__(self):
         # TODO: log to init/id tlparse after I add support for it
         log.info("ChromiumEventLogger initialized with id %s", self.id_)
 
+    def add_listener(self, listener: Callable[[Dict[str, Any]], None]) -> None:
```
cc @jamesjwu, since IIRC you added the `ChromiumEventLogger` code

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1509,12 +1523,24 @@ def _compiled_autograd_key(ctx):
             @staticmethod
             def forward(ctx, *deduped_flat_tensor_args):
+                if torch._functorch.config.aotd_debug_profile:
+                    get_chromium_event_logger().log_event_start(
```
did you sanity check that `get_chromium_event_logger().log_event_start()` is fast enough that it doesn't throw off our measurements? (especially since these runtime overhead benchmarks are in the range of micro-to-milliseconds, while most compile time benchmarks are seconds-to-minutes)

bdhirsh commented:
Does this subsume https://github.com/pytorch/pytorch/pull/135501 ?

IvanKobzarev commented:
> Does this subsume #135501 ?

I've reused the test logic and applied Edwards comment to move config in fuctorch/config, but dropped profile module that I started in that PR in favor of using chromium_logger. I will abandon/close #135501.

jamesjwu reviewed (commented):
Looks reasonable from chromium event side. Can you run a benchmark like nanogpt with training enabled and post the full chromium trace after adding this PR? 

Overall, I think I'd just like a little more documentation on what AOTD is somewhere near these events/config

IvanKobzarev commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -282,6 +284,12 @@ def _create_runtime_wrapper(
 
     def runtime_wrapper(args: List[Any]):
+        if torch._functorch.config.aotd_debug_profile:
+            get_chromium_event_logger().log_event_start(
```
Yes, here I want to measure runtime overhead of subclasses. Wrapping-unwrapping in wrappers etc.

IvanKobzarev commented:
> nanogpt

Do you know, if nanogpt uses subclasses?

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"


The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"


The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"


The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Subclasses profile logging"


The goal is:
Optional profiling of AOTD Runtime overhead, to chromium_logger.

Changes to ChromiumLogger:

Adding listener to be able to subscribe to events of the logger to:
1/ Verify in the tests that events are sent correctly
2/ Use the data from events for adhoc perf analysis of changes in AOTD

```
TORCH_TRACE="/var/tmp/tt" python test/functorch/test_aotdispatch.py -k test_aotd_debug_profile_overhead_logging
```
```
   SUBCLASS FW_aotd_overhead_avg:2264.533ns 51.080% of fwd_compiled_module_duration_avg_ns:4433.325ns fwd_runtime_wrapper_duration_avg_ns:6697.858ns
   SUBCLASS BW_aotd_overhead_avg:1449.392ns 19.426% of bwd_compiled_module_duration_avg:7461.050ns bwd_total_duration_avg:8910.442ns
NO_SUBCLASS FW_aotd_overhead_avg:2215.717ns 49.924% of fwd_compiled_module_duration_avg_ns:4438.175ns fwd_runtime_wrapper_duration_avg_ns:6653.892ns
NO_SUBCLASS BW_aotd_overhead_avg:1206.492ns 16.328% of bwd_compiled_module_duration_avg:7389.025ns bwd_total_duration_avg:8595.517ns
```



Perfetto trace:

![Screenshot 2024-09-24 at 16 22 54](https://github.com/user-attachments/assets/8a5293bb-6dbc-40e6-935a-5c5bd79ae717)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]
